src/lib/SaferNames/Inference.hs

```diff
@@ -363,6 +371,17 @@ instance HoistableE e => HoistableE (HoistedSolverState e) where
   freeVarsE (HoistedSolverState infVars defaults subst ab) =
     freeVarsE (Abs infVars (PairE (PairE defaults subst) ab))
 
+dceIfSolved :: (HoistableE e, SubstE AtomSubstVal e)
```

Comment:
I prefer these outlined functions because they have fewer variables in scope so I find it easier to reason about what they do. Obviously our tastes differ on this point. As we discussed, something that would satisfy us both would be a feature like `where` that lets you define a function locally within another function but *doesn't* give you access to the enclosing function's lambda-bound names.